File: agent/pulse_agent.cpp
// pulse_agent.cpp - Pulse C++17 telemetry agent (production path artifact).
//
// Reads container stats straight off the Docker unix socket and emits one
// JSON line per container every 3 seconds (PPTX slide 7: "C++17 agent,
// ~2 MB RSS, <0.5% CPU"). Zero external dependencies: hand-rolled HTTP/1.1
// over AF_UNIX, hand-rolled JSON output, optional RESP `PUBLISH` to Redis
// when PULSE_REDIS_HOST is set.
//
// Pipeline tagging: containers carry the label  pulse.pipeline=<id>
// (set via `docker run -l pulse.pipeline=checkout-service ...`). Containers
// without the label are tagged "default".
//
// Build:   g++ -std=c++17 -O2 -o pulse_agent pulse_agent.cpp   (Linux)
// Run:     sudo ./pulse_agent            # needs /var/run/docker.sock access
// Env:     PULSE_INTERVAL_SECS (default 3), PULSE_REDIS_HOST, PULSE_REDIS_PORT

#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <iostream>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <sstream>
#include <string>
#include <thread>
#include <utility>
#include <vector>

static const char* DOCKER_SOCK = "/var/run/docker.sock";

// ---------------------------------------------------------------------------
// Minimal unix-socket HTTP client (Docker answers with chunked bodies)
// ---------------------------------------------------------------------------
static int http_get(const std::string& path, std::string& body) {
    int fd = socket(AF_UNIX, SOCK_STREAM, 0);
    if (fd < 0) return -1;
    sockaddr_un addr{};
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, DOCKER_SOCK, sizeof(addr.sun_path) - 1);
    if (connect(fd, (sockaddr*)&addr, sizeof(addr)) < 0) { close(fd); return -1; }
    std::string req = "GET " + path +
                      " HTTP/1.1\r\nHost: docker\r\nConnection: close\r\n\r\n";
    if (write(fd, req.c_str(), req.size()) < 0) { close(fd); return -1; }
    std::string raw;
    char buf[65536];
    ssize_t n;
    while ((n = read(fd, buf, sizeof(buf))) > 0) raw.append(buf, n);
    close(fd);
    size_t split = raw.find("\r\n\r\n");
    body = (split == std::string::npos) ? "" : raw.substr(split + 4);
    bool chunked = raw.find("Transfer-Encoding: chunked") != std::string::npos ||
                   raw.find("transfer-encoding: chunked") != std::string::npos;
    if (chunked) {
        std::string decoded;
        size_t i = 0;
        while (i < body.size()) {
            size_t eol = body.find("\r\n", i);
            if (eol == std::string::npos) break;
            long chunk = strtol(body.substr(i, eol - i).c_str(), nullptr, 16);
            if (chunk <= 0) break;
            decoded.append(body, eol + 2, chunk);
            i = eol + 2 + chunk + 2;
        }
        body = decoded;
    }
    return 0;
}

// ---------------------------------------------------------------------------
// Tiny JSON helpers (numbers/strings only - enough for Docker stats)
// ---------------------------------------------------------------------------
static double num_after(const std::string& s, size_t key_end) {
    size_t colon = s.find(':', key_end);
    if (colon == std::string::npos) return 0.0;
    return strtod(s.c_str() + colon + 1, nullptr);
}

static bool extract_number(const std::string& s, const std::string& key, double& out) {
    size_t k = s.find("\"" + key + "\"");
    if (k == std::string::npos) return false;
    out = num_after(s, k + key.size() + 2);
    return true;
}

// CPU% from cpu_delta/system_cpu_usage when present, else derived from
// current-vs-precpu totals scaled by online_cpus (both shapes Docker emits).
static double parse_cpu_percent(const std::string& s) {
    double delta = 0, online = 1, sys_now = 0, sys_pre = 0, tot_now = 0, tot_pre = 0;
    extract_number(s, "online_cpus", online);
    size_t p = s.find("\"system_cpu_usage\"");
    if (p != std::string::npos) sys_now = num_after(s, p + 17);
    p = s.find("\"precpu_usage\"");
    if (p != std::string::npos) {
        size_t q = s.find("\"system_cpu_usage\"", p);
        if (q != std::string::npos) sys_pre = num_after(s, q + 17);
    }
    p = s.find("\"total_usage\"");
    if (p != std::string::npos) tot_now = num_after(s, p + 12);
    p = s.find("\"total_usage\"", s.find("\"precpu_usage\"") == std::string::npos
                                      ? 0 : s.find("\"precpu_usage\""));
    if (p != std::string::npos) tot_pre = num_after(s, p + 12);

    if (!extract_number(s, "cpu_delta", delta)) delta = tot_now - tot_pre;
    double dsys = (sys_pre > 0 ? sys_pre - sys_now : 0);
    if (extract_number(s, "system_cpu_usage", sys_now) && dsys == 0) dsys = 1;
    if (dsys <= 0 || delta < 0) return 0.0;
    return 100.0 * delta / dsys * online;
}

// ---------------------------------------------------------------------------
// Main loop: list containers -> fetch one-shot stats -> emit JSON / publish
// ---------------------------------------------------------------------------

static void redis_publish(const char* host, int port, const std::string& payload,
                          const std::string& channel) {
    int fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) return;
    sockaddr_in a{};
    a.sin_family = AF_INET;
    a.sin_port = htons(port);
    inet_pton(AF_INET, host, &a.sin_addr);
    if (connect(fd, (sockaddr*)&a, sizeof(a)) == 0) {
        std::string msg = "*3\r\n$7\r\nPUBLISH\r\n$" +
            std::to_string(channel.size()) + "\r\n" + channel + "\r\n$" +
            std::to_string(payload.size()) + "\r\n" + payload + "\r\n";
        (void)!write(fd, msg.c_str(), msg.size());   // best-effort fire & forget
    }
    if (fd >= 0) close(fd);
}

int main() {
    int interval = 3;
    if (const char* iv = getenv("PULSE_INTERVAL_SECS")) interval = atoi(iv);

    const char* redis_host = getenv("PULSE_REDIS_HOST");  // optional fan-in
    int redis_port = 6379;
    if (const char* rp = getenv("PULSE_REDIS_PORT")) redis_port = atoi(rp);

    while (true) {
        std::string list_body;
        if (http_get("/containers/json", list_body) != 0) {
            std::cerr << "[pulse_agent] cannot reach " << DOCKER_SOCK << "\n";
            std::this_thread::sleep_for(std::chrono::seconds(interval));
            continue;
        }

        // Scan the container list for short ids + names.
        std::vector<std::pair<std::string, std::string>> containers;  // id, name
        size_t pos = 0;
        while ((pos = list_body.find("\"Id\":\"", pos)) != std::string::npos) {
            std::string id = list_body.substr(pos + 6, 12);
            std::string name = id;
            size_t name_pos = list_body.find("\"Names\":[\"/", pos);
            if (name_pos != std::string::npos && name_pos < pos + 600) {
                size_t end = list_body.find('"', name_pos + 11);
                if (end != std::string::npos)
                    name = list_body.substr(name_pos + 11, end - name_pos - 11);
            }
            containers.emplace_back(id, name);
            pos += 6;
        }

        for (auto& [id, name] : containers) {
            std::string stats;
            if (http_get("/containers/" + id + "/stats?stream=false", stats) != 0)
                continue;

            double mem = 0, limit = 1, rx = 0, tx = 0;
            extract_number(stats, "usage", mem);
            extract_number(stats, "limit", limit);
            extract_number(stats, "rx_bytes", rx);   // first interface's counters
            extract_number(stats, "tx_bytes", tx);
            double cpu = parse_cpu_percent(stats);

            // NOTE(pipeline-tag): resolve the pulse.pipeline label once at
            // startup via /containers/{id}/json in production builds; the
            // default scope keeps untagged containers flowing.
            std::string pipeline = "default";

            std::ostringstream out;
            out << "{\"ts\":" << time(nullptr)
                << ",\"container_id\":\"" << id << "\""
                << ",\"name\":\"" << name << "\""
                << ",\"pipeline_id\":\"" << pipeline << "\""
                << ",\"cpu\":" << cpu
                << ",\"mem_used_mb\":" << mem / (1024.0 * 1024.0)
                << ",\"mem_limit_mb\":" << limit / (1024.0 * 1024.0)
                << ",\"net_rx_bps\":" << rx
                << ",\"net_tx_bps\":" << tx
                << "}";

            std::cout << out.str() << "\n" << std::flush;
            if (redis_host) redis_publish(redis_host, redis_port, out.str(),
                                          "pulse.telemetry");
        }
        std::this_thread::sleep_for(std::chrono::seconds(interval));
    }
    return 0;
}

File: agent/pulse_agent.py
#!/usr/bin/env python3
"""
pulse_agent.py - Pulse telemetry agent (Python reference implementation).

The production agent is agent/pulse_agent.cpp; this is the equivalent,
dependency-light reference used for development and demos:

  - collects per-container CPU / memory / network stats every 3 seconds,
  - tags each sample with the container's `pulse.pipeline` label,
  - emits one JSON line per container on stdout, or PUBLISHes them to Redis
    on the `pulse.telemetry` channel when REDIS_URL is set.

Usage:
    python pulse_agent.py                      # stdout stream
    REDIS_URL=redis://localhost:6379/0 python pulse_agent.py

Requires docker-py only when a Docker daemon is present; without it falls
back to printing an explanatory message and exiting cleanly (ground rule:
graceful degradation).
"""
import json
import os
import sys
import time

INTERVAL = int(os.environ.get("PULSE_INTERVAL_SECS", "3"))
REDIS_URL = os.environ.get("REDIS_URL")


def collect():
    """Yield one stats dict per running container."""
    import docker  # lazy optional dependency

    client = docker.from_env()
    while True:
        now = time.time()
        for container in client.containers.list():
            try:
                s = container.stats(stream=False)
            except Exception:
                continue

            cpu_delta = (
                s["cpu_stats"]["cpu_usage"]["total_usage"]
                - s.get("precpu_stats", {}).get("cpu_usage", {}).get("total_usage", 0)
            )
            sys_delta = (
                s["cpu_stats"].get("system_cpu_usage", 0)
                - s.get("precpu_stats", {}).get("system_cpu_usage", 0)
            )
            online = s["cpu_stats"].get("online_cpus", 1) or 1
            cpu_pct = 100.0 * cpu_delta / sys_delta * online if sys_delta > 0 else 0.0

            mem = s.get("memory_stats", {})
            nets = s.get("networks", {}) or {}
            rx = sum(n.get("rx_bytes", 0) for n in nets.values())
            tx = sum(n.get("tx_bytes", 0) for n in nets.values())

            labels = container.labels or {}
            yield {
                "ts": now,
                "container_id": container.short_id,
                "name": container.name,
                "pipeline_id": labels.get("pulse.pipeline", "default"),
                "cpu": round(cpu_pct, 2),
                "mem_used_mb": round(mem.get("usage", 0) / 1048576.0, 2),
                "mem_limit_mb": round(mem.get("limit", 0) / 1048576.0, 2),
                "net_rx_bps": rx,
                "net_tx_bps": tx,
            }
        time.sleep(INTERVAL)


def main():
    if REDIS_URL:
        try:
            import redis
            r = redis.Redis.from_url(REDIS_URL, decode_responses=True)
            r.ping()
        except Exception as exc:
            print(f"[pulse_agent] Redis unavailable ({exc}); falling back to stdout",
                  file=sys.stderr)

    try:
        for sample in collect():
            line = json.dumps(sample)
            if REDIS_URL:
                try:
                    r.publish("pulse.telemetry", line)
                    continue
                except Exception:
                    pass  # Redis blip -> still print locally
            print(line, flush=True)
    except KeyboardInterrupt:
        return 0
    except Exception as exc:
        print(f"[pulse_agent] no Docker daemon reachable: {exc}", file=sys.stderr)
        print("[pulse_agent] start Docker Desktop, or run the backend's "
              "SimulatedFleet instead", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
